File: clients/go/txn.go
package volant

import (
	"fmt"

	"github.com/volant-mq/volant/clients/go/codec"
)

// TxnOffset is one (topic, partition, offset) triple queued by
// [TransactionalProducer.AddOffsets]. Sent on Commit as a
// [codec.TxnOffsetCommit] (empty metadata).
type TxnOffset struct {
	Topic     string
	Partition uint32
	Offset    uint64
}

// Re-exported EndTxn row types (v0.57).
type (
	TxnOffsetCommit  = codec.TxnOffsetCommit
	TxnProduceResult = codec.TxnProduceResult
)

// TransactionalProducer is a thin wrapper around Client BeginTxn / EndTxn
// (v0.63). Matches crates/volant-client/src/txn.rs. Native opcodes 50–53
// only; not Kafka transactions. Produce is write-through.
type TransactionalProducer struct {
	client  *Client
	pending []codec.TxnOffsetCommit
	open    bool
}

// NewTransactionalProducer wraps a Client that already has a non-empty
// transactional_id (same check as BeginTransaction).
func NewTransactionalProducer(c *Client) (*TransactionalProducer, error) {
	if c == nil || c.transactionalID == "" {
		return nil, fmt.Errorf("transactional_id not configured")
	}
	return &TransactionalProducer{client: c}, nil
}

// Client returns the underlying Client.
func (p *TransactionalProducer) Client() *Client {
	return p.client
}

// Begin opens a native transaction. Double-begin while open is an error.
func (p *TransactionalProducer) Begin() error {
	if p.open {
		return fmt.Errorf("transaction already open")
	}
	if err := p.client.BeginTransaction(); err != nil {
		return err
	}
	p.pending = p.pending[:0]
	p.open = true
	return nil
}

// Produce delegates to [Client.Produce] (write-through).
func (p *TransactionalProducer) Produce(topic string, partition int, key, value []byte) (int64, error) {
	return p.client.Produce(topic, partition, key, value)
}

// AddOffsets queues group offsets locally. Nothing is sent until Commit.
func (p *TransactionalProducer) AddOffsets(groupID string, offsets []TxnOffset) {
	for _, o := range offsets {
		p.pending = append(p.pending, codec.TxnOffsetCommit{
			GroupID:   groupID,
			Topic:     o.Topic,
			Partition: o.Partition,
			Offset:    o.Offset,
		})
	}
}

// AddOffset queues a single topic/partition/offset triple.
func (p *TransactionalProducer) AddOffset(groupID, topic string, partition uint32, offset uint64) {
	p.AddOffsets(groupID, []TxnOffset{{Topic: topic, Partition: partition, Offset: offset}})
}

// Commit sends EndTxn committed=1 with the queued offsets.
func (p *TransactionalProducer) Commit() ([]codec.TxnProduceResult, error) {
	if !p.open {
		return nil, fmt.Errorf("transaction is not open")
	}
	offsets := p.pending
	p.pending = nil
	results, err := p.client.CommitTransaction(offsets)
	if err != nil {
		return nil, err
	}
	p.open = false
	return results, nil
}

// Abort clears the offset queue and sends EndTxn committed=0.
func (p *TransactionalProducer) Abort() error {
	if !p.open {
		return fmt.Errorf("transaction is not open")
	}
	p.pending = nil
	if err := p.client.AbortTransaction(); err != nil {
		return err
	}
	p.open = false
	return nil
}

// IsOpen reports whether a transaction is open locally.
func (p *TransactionalProducer) IsOpen() bool {
	return p.open
}

File: clients/python/src/volant/txn.py
"""Transactional producer helper (v0.63).

Thin wrapper around the v0.57 Client BeginTxn / EndTxn RPCs. Matches
``crates/volant-client/src/txn.rs``. Native opcodes 50–53 only; not
Kafka transactions.
"""

from __future__ import annotations

from typing import Iterable, Optional, Union

from .client import Client, ProduceResult
from .codec import ProduceMessage, TxnOffsetCommit, TxnProduceResult


class TransactionalProducer:
    """begin → produce* / add_offsets* → commit or abort.

    ``add_offsets`` queues locally; nothing is sent until ``commit``.
    Produce is write-through (same as the Rust helper).
    """

    def __init__(self, client: Client) -> None:
        if client is None or not client.transactional_id:
            raise ValueError("transactional_id not configured")
        self._client = client
        self._pending: list[TxnOffsetCommit] = []
        self._open = False

    @property
    def client(self) -> Client:
        return self._client

    def begin(self) -> None:
        if self._open:
            raise ValueError("transaction already open")
        self._client.begin_transaction()
        self._pending.clear()
        self._open = True

    def produce(
        self,
        topic: str,
        partition: int,
        value: Optional[bytes] = None,
        *,
        key: Optional[bytes] = None,
        messages: Optional[Iterable[Union[bytes, ProduceMessage]]] = None,
        acks: Optional[int] = None,
        timestamp_ms: int = -1,
        headers: Optional[list[tuple[str, bytes]]] = None,
    ) -> ProduceResult:
        return self._client.produce(
            topic,
            partition,
            value=value,
            key=key,
            messages=messages,
            acks=acks,
            timestamp_ms=timestamp_ms,
            headers=headers,
        )

    def add_offsets(
        self,
        group_id: str,
        entries: Iterable[tuple[str, int, int]],
    ) -> None:
        """Queue group offsets to commit atomically with EndTxn.

        ``entries`` are ``(topic, partition, offset)`` triples. Nothing is
        sent until :meth:`commit`.
        """
        for topic, partition, offset in entries:
            self._pending.append(
                TxnOffsetCommit(
                    group_id=group_id,
                    topic=topic,
                    partition=partition,
                    offset=offset,
                    metadata="",
                )
            )

    def commit(self) -> list[TxnProduceResult]:
        if not self._open:
            raise ValueError("transaction is not open")
        offsets = self._pending
        self._pending = []
        results = self._client.commit_transaction(offsets=offsets)
        self._open = False
        return results

    def abort(self) -> None:
        if not self._open:
            raise ValueError("transaction is not open")
        self._pending.clear()
        self._client.abort_transaction()
        self._open = False

    def is_open(self) -> bool:
        return self._open
